packages/language-server/src/plugins/typescript/serviceCache.ts

```diff
@@ -0,0 +1,93 @@
+// abstracting the typescript-auto-import-cache package to support our use case
+
+import {
+    ProjectService,
+    createProjectService as createProjectService50
+} from 'typescript-auto-import-cache/out/5_0/projectService';
+import { createProject as createProject50 } from 'typescript-auto-import-cache/out/5_0/project';
+import { createProject as createProject53 } from 'typescript-auto-import-cache/out/5_3/project';
+import ts from 'typescript';
+import { ExportInfoMap } from 'typescript-auto-import-cache/out/5_0/exportInfoMap';
+import { ModuleSpecifierCache } from 'typescript-auto-import-cache/out/5_0/moduleSpecifierCache';
+import { SymlinkCache } from 'typescript-auto-import-cache/out/5_0/symlinkCache';
+import { ProjectPackageJsonInfo } from 'typescript-auto-import-cache/out/5_0/packageJsonCache';
+
+export { ProjectService };
+
+declare module 'typescript' {
+    interface LanguageServiceHost {
+        /** @internal */ getCachedExportInfoMap?(): ExportInfoMap;
+        /** @internal */ getModuleSpecifierCache?(): ModuleSpecifierCache;
+        /** @internal */ getGlobalTypingsCacheLocation?(): string | undefined;
+        /** @internal */ getSymlinkCache?(files: readonly ts.SourceFile[]): SymlinkCache;
+        /** @internal */ getPackageJsonsVisibleToFile?(
+            fileName: string,
+            rootDir?: string
+        ): readonly ProjectPackageJsonInfo[];
+        /** @internal */ getPackageJsonAutoImportProvider?(): ts.Program | undefined;
+        /** @internal */ useSourceOfProjectReferenceRedirect?(): boolean;
+    }
+}
+
+export function createProjectService(
+    system: ts.System,
+    hostConfiguration: {
+        preferences: ts.UserPreferences;
+    }
+) {
+    const version = ts.version.split('.');
+    const major = parseInt(version[0]);
+
+    if (major < 5) {
+        return undefined;
+    }
+
+    const projectService = createProjectService50(
+        ts,
+        system,
+        system.getCurrentDirectory(),
+        hostConfiguration,
+        ts.LanguageServiceMode.Semantic
+    );
+
+    return projectService;
+}
+
+export function createProject(
+    host: ts.LanguageServiceHost,
+    createLanguageService: (host: ts.LanguageServiceHost) => ts.LanguageService,
+    options: {
+        projectService: ProjectService;
+        compilerOptions: ts.CompilerOptions;
+        currentDirectory: string;
+    }
+) {
+    const version = ts.version.split('.');
+    const major = parseInt(version[0]);
+    const minor = parseInt(version[1]);
+
+    if (major < 5) {
+        return undefined;
+    }
+
+    const factory = minor < 3 ? createProject50 : createProject53;
```

Comment:
This makes me a bit nervous that in the future we potentially need to adjust this for new TS versions - but I guess there's no other way around this since this is using internal methods. Ideally the cache becomes part of the TS service at some point...

Reply:
Yeah. This might break in the future. Maybe it'll be better to let the `typescript-auto-import-cache` handle it, but there is the integration problem I mentioned above. I talked with @/johnsoncodehk, but he's not sure if the new API we added now in `typescript-auto-import-cache` can be available later since the potential changes in the future and with Volar 2.

Reply:
Just to be sure: This doesn't mean that `typescript-auto-import-cache` isn't going away at some point, it's just that the APIs we need on top of it cannot be built into the package itself right now because of the potential Volar 2 changes.

Reply:
Yeah. Not saying the package is going away. I checked the chat history. What he said is that he might "bundle" the code responsible for the cache.